### ISSUE TITLE
Feature request: Full support for Anonymous / Singleton resource

### ISSUE THREAD

**Issue by slavo2** [api-platform/api-platform#1198]
# Proposal: Support for Anonymous resources

So far there are only 3 kinds of operation allowed:
- item operations (these expects url contains an id)
- collection operations (these expects result is a collection)
- subresource operations

It is common need to define and document an endpoint, which have neither of these attributes,
but could be (in given context - e.g. for given authenticated User) considered a singleton.
e.g. '/me' or '/user/current' or '/auth/login' or ' /settings' etc.
It's structure might be well defined, it is a single item (not a collection) it has no (or does need no) ID.
Currently it is possible to implement this with custom item operation with a small hack, e.g. to specify:

```php
/**
 * @ApiResource(
 *     itemOperations={
 *          "get_current"={
 *              "method"="GET",
 *              "path"="/user/current",
 *              "controller"=GetCurrentUserController::class,
 *              "swagger_context"={
 *                  "parameters" = {}
 *              },
 *              "openapi_context"={
 *                  "parameters" = {}
 *              },
 *              "defaults"={"_api_receive"=false}
 *          }
 *     }
 * )
 */
class User 
{
}
```

but it might be bit of a hack.

It would be better, that ApiPlatform would allow defining ID-less / Anonymous / Singleton / Singular (pick your term :) ) resources standard way.

#### References:
(Official support not yet @teohhanhui in
https://github.com/api-platform/api-platform/issues/199#issuecomment-270881649
on 6/1/2017)

**Comment by teohhanhui** [comment]
I don't think this is the right approach. A resource must have a dereferencable IRI, by definition.

**Comment by teohhanhui** [comment]
But we should probably rethink our arbitrary distinction of collection / item operations and the assumptions around that.

**Comment by slavo2** [comment]
@teohhanhui But it is a common use. If you want to (write and) document endpoints connected to authentication (where the need to retrieve currently logged in user details (without specifying the user ID in the path) is quite common I suppose), it would be handy if you can define your own operations, where the framework / library would not be standing in the way - supposing you have to have an ID in the path for working with individual resources.

**Comment by slavo2** [comment]
> I don't think this is the right approach. A resource must have a dereferencable IRI, by definition.

Well, it can have an IRI and not have an ID in a sense... like '/user/current' might be considered an IRI ;)
I admit its meaning is context-bound, but so are collections, implicitly filtered by currently logged in user. ( https://api-platform.com/docs/core/security/ -> second-last paragraph)

**Comment by teohhanhui** [comment]
The emphasis is on **dereferencable** IRI. :see_no_evil:

**Comment by slavo2** [comment]
Well, but how do you implement an API with authentication for a frontend client library, which expect that you have some endpoints with login, current user and optionally logout meaning.

Note if one would like to implement it in API platform and have the swagger/openapi documentation have automatically generated from models representing objects being sent over (and not needing to add in some documentation part manually)

e.g. https://auth.nuxtjs.org/schemes/local.html#options

**Comment by teohhanhui** [comment]
You can see an example in https://api-platform.com/docs/core/jwt/

It should not be an API resource, so you should document it manually indeed. Unless you have a better suggestion... :smile:

**Comment by slavo2** [comment]
> You can see an example in https://api-platform.com/docs/core/jwt/
> 
> It should not be an API resource, so you should document it manually indeed. Unless you have a better suggestion... 😄

Well, exactly, that is the problem. If one follows the documentation on how to add JWT still needs to add documentation manually. (It would be nice however, if the package would provide it automatically or the documentation on JWT package would provide an example how to do it properly...)

My suggestion is just above :-)

**Comment by slavo2** [comment]
Btw. I would like to ask what is the ApiPlatform project relation to  

https://medium.com/@ro0NL/domain-driven-design-projections-in-practice-with-api-platform-and-elasticsearch-c785ed6d660b ?

Seems that https://github.com/msgphp/symfony-demo-app might be using some interesting approach...

**Comment by soyuka** [comment]
Closing per #2987. Questions are best asked on [GitHub Discussions](https://github.com/api-platform/core/discussions) or [Stack Overflow (api-platform tag)](https://stackoverflow.com/questions/tagged/api-platform.com). The framework now lives at `api-platform/core`.
